modules/drivers/oracle/src/metabase/driver/oracle.clj

```diff
@@ -599,22 +599,26 @@
   (fn []
     (.getString rs i)))
 
+(defn- ^Connection rs->conn [^ResultSet rs]
+  (let [^C3P0ProxyConnection proxy-conn (.. rs getStatement getConnection)]
+    (.unwrap proxy-conn OracleConnection)))
+
+;; Oracle `TIMESTAMPTZ` or `TIMESTAMPLTZ` types can have either a zone offset *or* a zone ID; you could fetch either
+;; `OffsetDateTime` or `ZonedDateTime` using `.getObject`, but fetching the wrong type will result in an Exception,
+;; meaning we have try both and wrap the first in a try-catch. As far as I know there's no way to tell whether the
+;; value has a zone offset or ID without first fetching a `TIMESTAMPTZ` object. So to avoid the try-catch we can
+;; fetch the `TIMESTAMPTZ` or `TIMESTAMPLTZ` and use `.timestampValue` instead.
 (defmethod sql-jdbc.execute/read-column-thunk [:oracle OracleTypes/TIMESTAMPTZ]
   [_driver ^ResultSet rs _rsmeta ^Integer i]
-  ;; Oracle `TIMESTAMPTZ` types can have either a zone offset *or* a zone ID; you could fetch either `OffsetDateTime`
-  ;; or `ZonedDateTime` using `.getObject`, but fetching the wrong type will result in an Exception, meaning we have
-  ;; try both and wrap the first in a try-catch. As far as I know there's now way to tell whether the value has a zone
-  ;; offset or ID without first fetching a `TIMESTAMPTZ` object. So to avoid the try-catch we can fetch the
-  ;; `TIMESTAMPTZ` and use `.offsetDateTimeValue` instead.
   (fn []
     (when-let [^TIMESTAMPTZ t (.getObject rs i TIMESTAMPTZ)]
-      (let [^C3P0ProxyConnection proxy-conn (.. rs getStatement getConnection)
-            conn                            (.unwrap proxy-conn OracleConnection)]
-        ;; TIMEZONE FIXME - we need to warn if the Oracle JDBC driver is `ojdbc7.jar`, which probably won't have this
-        ;; method
-        ;;
-        ;; I think we can call `(oracle.jdbc.OracleDriver/getJDBCVersion)` and check whether it returns 4.2+
-        (.offsetDateTimeValue t conn)))))
+      (t/offset-date-time (.timestampValue t (rs->conn rs)) (t/zone-offset 0)))))
```

Comment:
I think we are loosing information on actual zone offset/zone as stored in the database here. But I believe that _may_ not be a problem as we do format results in `format-rows` middleware.

Reply:
> I suppose the old implementation returned OffsetDateTime with the actual offset

Yes.

These are good questions, and I'm looking for your input to confirm. The offset was untested for Oracle. I decided that `read-column-thunk` should return an `OffsetDateTime` in UTC because this is what all the other drivers do in [`metabase.query-processor-test.timezones-test/set-timezone-drivers`](https://github.com/metabase/metabase/blob/f23d75b06c438b3692cf73d322b9792937cd66a2/test/metabase/query_processor_test/timezones_test.clj#L31-L36), which includes Postgres and MySQL.

@camsaul can you confirm the desired behaviour here?

Reply:
I agree, this behavior seems pretty questionable. `.offsetDateTimeValue()` seems way saner, what's the reason for changing this?

According to

https://docs.oracle.com/en/database/oracle/oracle-database/19/nlspg/datetime-data-types-and-time-zone-support.html#GUID-3F1C388E-C651-43D5-ADBC-1A49E5C2CA05

> TIMESTAMP WITH LOCAL TIME ZONE is another variant of TIMESTAMP. It differs from TIMESTAMP WITH TIME ZONE as follows: data stored in the database is normalized to the database time zone, and the time zone offset is not stored as part of the column data. When users retrieve the data, Oracle Database returns it in the users' local session time zone. The time zone offset is the difference (in hours and minutes) between local time and UTC (Coordinated Universal Time, formerly Greenwich Mean Time).

We set the session timezone to whatever the `report-timezone` setting is, which means `TIMESTAMPLTZ` values should come back in the `report-timezone`.

According to

https://docs.oracle.com/en/database/oracle/oracle-database/21/jajdb/oracle/sql/TIMESTAMPLTZ.html#timestampValue(java.sql.Connection)

`.toTimestamp()` 

> Calls toTimestamp to convert internal Oracle TIMESTAMPLTZ to a Java Timestamp.

JavaTimestamps can have an attached `java.util.Calendar` specifying the time zone, in practice this doesn't seem to be done consistently so we can't really be sure whether the timestamp is coming back in UTC or the session timezone and has time zone information or not

Reply:
> what's the reason for changing this?

I mentioned above that I made the change so that Oracle was consistent with other drivers behaviour. That's the only reason: consistency, which is why I'm happy that you're questioning it, and I'm hoping we can write a test that enforces the desired behaviour across all drivers. Oracle was the only driver failing this test before the change:

https://github.com/metabase/metabase/blob/5dfe03915caae54382645814d4802686cc2df4a5/test/metabase/query_processor_test/timezones_test.clj#L245-L253

Reply:
Sorry, I missed your comment above an was replying to Libor. 

The offset that a `java.time.OffsetDateTime` comes back from JDBC isn't particularly important as long as it refers to a particular moment in time, because the `format-rows` middleware will adjust it so it has the `report-timezone` offset.

So either `2024-06-20T16:05:00-08:00` (PDT) or `2024-06-21T00:05:00Z` (UTC) would be ok as results for `read-column-thunk` since they refer to the same moment in time and `format-rows` would adjust them both to `2024-06-20T16:05:00-08:00` is `report-timezone` is `US/Pacific`.

That being said the test above doesn't really prove that the old behavior was incorrect, there might be a bug in the test data loading code for Oracle or something else might be going on. That's another reason I suggested writing a Oracle-specific test that ran a raw SQL query returning a `TIMESTAMP WITH LOCAL TIME ZONE` literal, because then we're literally just testing that we read a value out correctly... see `metabase.driver.athena-test/read-time-and-timestamp-with-time-zone-columns-test` for an example of this

Reply:
Oracle and `TIMESTAMP WITH LOCAL TIME ZONE` is kind of a weird special case to be honest

Reply:
I wrote this test and it's working correctly for me with the existing code which makes me think maybe something is off with our test data loading code for Oracle or something. Most of our other DBs don't actually store zone info, Postgres for example normalizes everything to UTC.

```clj
(deftest read-timestamp-with-local-time-zone-test
  (mt/test-driver :oracle
    (testing "We should return TIMESTAMP WITH LOCAL TIME ZONE columns correctly"
      ;; these both come back as `java.sql.type/VARCHAR` for some wacko reason from the JDBC driver, so let's make sure
      ;; we have code in place to work around that.
      (let [sql          "SELECT timestamp '2024-06-20 16:05:00 -07:00' AS timestamp_tz FROM dual"
            query        (-> (mt/native-query {:query sql})
                             (assoc-in [:middleware :format-rows?] false))]
        ;; value should come back from driver with original zone info, regardless of report timezone since we're not
        ;; applying format-rows middleware
        (doseq [report-timezone ["UTC" "US/Pacific"]]
          (mt/with-temporary-setting-values [report-timezone report-timezone]
            (mt/with-native-query-testing-context query
              (is (= [[#t "2024-06-20T16:05-07:00"]]
                     (mt/rows (qp/process-query query)))))))))))
```

Reply:
So I think what you're doing does work as well as long as we're applying the `format-rows` middleware, if I change the implementation for `TIMESTAMPTZ` to

```clj
(t/offset-date-time (.timestampValue t conn) (t/zone-offset 0))
```

and then run the test again it returns 

```clj
#t "2024-06-20T23:05Z"
```

which is the same instant in time, however it is losing the offset info that Oracle has.

Reply:
> TIMESTAMPLTZ values should come back in the report-timezone.

I'd question that because it's not what we do for other drivers, per the test above. Is there a reason to do it differently for oracle?

> JavaTimestamps can have an attached java.util.Calendar specifying the time zone

The returned type is `java.sql.Timestamp`, which definitely doesn't have a `java.util.Calendar` attached. I double checked the [java docs](https://docs.oracle.com/javase/8/docs/api/java/sql/Timestamp.html) and it has no notion of a time zone.

> we can't really be sure whether the timestamp is coming back in UTC or the session timezone

I definitely agree with this though, the javadocs don't specify what time zone the Timestamp value is in. I assume it's UTC regardless, but given it's undocumented I wouldn't trust it.

In [this commit](a4d15ba6e9084b97ca686600c80d96a376c32665), I've now reverted back to using `.offsetDateTimeValue` for `TIMESTAMPTZ`, and now use the `.timestampValue` method that takes a `java.util.Calendar` argument so we can guarantee that the returned Timestamp is in UTC.

Reply:
Yeah ok this all makes perfect sense, because Postgres and the like normalize `TIMESTAMP WITH TIME ZONE` to UTC the value that gets returned with `:format-rows? false` should come back as UTC, but for Oracle since it actually records offset I would expect the correct answer to preserve the original offset. So the tests there are a little broken but we haven't ran into problems with it since we didn't test against Oracle in the past. 

So you should keep the old `.offsetDateTimeValue()` implementation and update the expected rows for that test. You can add a new feature called something like `::timestamp-tz-preserves-offset`, default `false`, but `true` for Oracle, and then drivers that have that flag set can expect the rows that preserve offset while others get the rows normalized to UTC

Reply:
RE `java.sql.Timestamp` "definitely" not having a notion of zone offset, I promise you it does, it inherits from `java.util.Date` which can have a `java.util.Calendar`. Check out this method for example

https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#getTimezoneOffset--

That being said it's not used consistently by different JDBC drivers and `java.util.Date`/`java.sql.Date`/`java.sql.Timestamp` are all hot pieces of trash and we should absolutely not use them if we want stuff to work right. It's better to splice temporal values directly into the query as literals and read values out as strings then to use `java.util.Date` and friends

Reply:
> I assume it's UTC regardless

Honestly in my experience this is a very dangerous assumption to make you can't trust anyone to ever do something sane with a `java.sql.Timestamp`, you should just not use them

Reply:
Sure, I can use `offsetDateTimeValue` for `TIMESTAMPTZ` and preserve the original offset. I'm curious, does it actually make a difference though? Isn't row formatting applied in all cases?

But for `TIMESTAMPLTZ`, `offsetDateTimeValue` doesn't seem to exist for the driver version (despite the docs). So I still have to use `.timestampValue` with a UTC calendar as I'm currently doing [here](https://github.com/metabase/metabase/blob/baf86f7da3824555cb66ac948295415f811cd914/modules/drivers/oracle/src/metabase/driver/oracle.clj#L622).

Reply:
RE Timestamps having Calendars, you're right and I totally got it backwards. Reading Timestamps from JDBC ResultSets can take Calendars as a parameter as you've noticed above. If you don't specify a `Calendar` it might use the system calendar/timezone, it might use the database timezone, it might use UTC, who knows. Explicitly specifying it is the way to go

Reply:
> Isn't row formatting applied in all cases?

I think maybe we're skipping it for Excel downloads or something also if `report-timezone` is unset I don't think we apply it timezone conversion logic, or at least one point we didn't. Either way I think it's nice to not throw out info until we get to that point in case we want to make it more sophisticated in the future or whatever

Reply:
`offsetDateTimeValue` seems to exist for me

```clj
(get-in (clojure.datafy/datafy oracle.sql.TIMESTAMPLTZ) [:members 'offsetDateTimeValue])
[#clojure.reflect.Method{:name offsetDateTimeValue,
                         :return-type java.time.OffsetDateTime,
                         :declaring-class oracle.sql.TIMESTAMPLTZ,
                         :parameter-types [java.sql.Connection],
                         :exception-types [java.sql.SQLException],
                         :flags #{:public}}] 
```

Reply:
Cool, I'll leave the non-test changes as is and ping you again when I've updated the driver tests.

Reply:
> offsetDateTimeValue seems to exist for me

Yes, I get the same. Can you call it though?

Reply:
Uh, now it's working for me too. Not sure what I was doing wrong, maybe I was trying `toOffsetDateTime` instead.​

Reply:
Yeah it does work for me but this test I wrote for it is failing, it's coming back as `-8:00` for `US/Pacific` rather than `-07:00` which is worrisome

```clj
(defmethod sql-jdbc.execute/read-column-thunk [:oracle OracleTypes/TIMESTAMPLTZ]
  [_driver ^ResultSet rs _rsmeta ^Integer i]
  (fn []
    (when-let [^TIMESTAMPLTZ t (.getObject rs i TIMESTAMPLTZ)]
      (let [^C3P0ProxyConnection proxy-conn (.. rs getStatement getConnection)
            ^Connection conn                (.unwrap proxy-conn OracleConnection)]
        (.offsetDateTimeValue t conn)))))
```

```clj
(deftest read-timestamp-with-local-time-zone-test
  (mt/test-driver :oracle
    (testing "We should return TIMESTAMP WITH LOCAL TIME ZONE columns correctly"
      (let [sql          "SELECT cast(timestamp '2024-06-20 16:05:00 -07:00' AS timestamp with local time zone) AS timestamp_ltz FROM dual"
            query        (-> (mt/native-query {:query sql})
                             (assoc-in [:middleware :format-rows?] false))]
        ;; value should come back from driver normalized to the report timezone
        (mt/with-temporary-setting-values [report-timezone "US/Pacific"]
          (mt/with-native-query-testing-context query
            (is (= [[#t "2024-06-20T16:05-07:00"]]
                   (mt/rows (qp/process-query query))))))
        (mt/with-temporary-setting-values [report-timezone "UTC"]
          (mt/with-native-query-testing-context query
            (is (= [[(t/offset-date-time #t "2024-06-20T23:05Z")]]
                   (mt/rows (qp/process-query query))))))))))
```

Reply:
I'm sure something really dumb is going on here, maybe Oracle is being super dumb and not applying DST or something. `SESSIONTIMEZONE` does appear to be correctly set to `US/Pacific`

Reply:
Investigating rn

Reply:
Oh cool this is just great Oracle is not broken at all 

```clj
(defn x []
  (mt/with-driver :oracle
    (sql-jdbc.execute/do-with-connection-with-options
     :oracle
     (mt/db)
     {:session-timezone "US/Pacific"}
     (fn [^com.mchange.v2.c3p0.C3P0ProxyConnection proxy-conn]
       (let [^oracle.jdbc.OracleConnection conn (.unwrap proxy-conn oracle.jdbc.OracleConnection)
             t (oracle.sql.TIMESTAMPLTZ. conn "2024-06-20 17:28:00 -07:00")]
         (.offsetDateTimeValue t conn))))))
(x)
;; => #t "2024-06-20T17:28-08:00"
```

Reply:
I remember now, I didn't use `timestampValue` because `offsetDateTimeValue` didn't exist, I used it because `offsetDateTimeValue` returns the incorrect results with DST when the value is stored in the DB.

This test fails with `offsetDateTimeValue`, returning a timestamp that is off by an hour:
```
(deftest sql-datetime-timezone-handling-test
  (mt/test-drivers (filter #(isa? driver/hierarchy % :sql)
                           (mt/normal-drivers-with-feature :set-timezone))
    (mt/dataset dt-attempted-murders
      (doseq [timezone [#_"UTC" "US/Pacific" #_"US/Eastern" #_"Asia/Hong_Kong"]]
        (mt/with-temporary-setting-values [report-timezone timezone]
          (let [expected        {:datetime_ltz (t/offset-date-time #t "2019-11-01T00:23:18.331-07:00" (t/zone-offset -7))
                                 :datetime_tz  (t/offset-date-time #t "2019-11-01T00:23:18.331-07:00" (t/zone-offset -7))}
                supported-type? {:datetime_ltz (driver.u/supports? driver/*driver* ::datetime-with-local-tz (mt/db))
                                 :datetime_tz  (driver.u/supports? driver/*driver* ::datetime-with-tz (mt/db))}
                actual          (select-keys (dt-attempts) (keys expected))
                keep-supported-types #(m/filter-keys supported-type? %)]
            (is (= (keep-supported-types expected)
                   (keep-supported-types actual)))))))))

; expected: {:datetime_ltz #t "2019-11-01T00:23:18.331-07:00", :datetime_tz #t "2019-11-01T00:23:18.331-07:00"}
;   actual: {:datetime_ltz #t "2019-11-01T00:23:18.331-08:00", :datetime_tz #t "2019-11-01T00:23:18.331-07:00"}
;     diff: - {:datetime_ltz #t "2019-11-01T00:23:18.331-07:00"}
;           + {:datetime_ltz #t "2019-11-01T00:23:18.331-08:00"}
```
It's fine when using `timestampValue` though.

Curiously it doesn't fail when using `select`. This passes:
```
(deftest read-timestamp-with-local-time-zone-test
  (mt/test-driver :oracle
    (testing "We should return TIMESTAMP WITH LOCAL TIME ZONE columns correctly"
      ;; these both come back as `java.sql.type/VARCHAR` for some wacko reason from the JDBC driver, so let's make sure
      ;; we have code in place to work around that.
      (let [sql   "SELECT timestamp '2019-11-01 00:23:18 -07:00' AS timestamp_tz FROM dual"
            query (-> (mt/native-query {:query sql})
                      (assoc-in [:middleware :format-rows?] false))]
        ;; value should come back from driver with original zone info, regardless of report timezone since we're not
        ;; applying format-rows middleware
        (doseq [report-timezone ["UTC" "US/Pacific"]]
          (mt/with-temporary-setting-values [report-timezone report-timezone]
            (mt/with-native-query-testing-context query
              (is (= [[#t "2019-11-01T00:23:18-07:00"]]
                     (mt/rows (qp/process-query query)))))))))))
```

Reply:
Hah, I didn't see your above comments. Looks like you found the same bug.

Reply:
Anyway, seems like we can use `timestampValue` to get around this issue for now. wdyt?

Reply:
ok given that `.offsetDateTimeValue()` seems to be broken here I think maybe just calling `stringValue` and parsing it isn't the worst idea in the world.

This gives us the actual correct value and we're not at the whim of broken implementations of `offsetDateTimeValue()` or `timestampValue()` or whatever.

If the Oracle JDBC people can't get `offsetDateTimeValue()` right I don't think we should trust the other stuff either. String however is probably pretty hard to get wrong

```clj
(defn x []
  (mt/with-driver :oracle
    (sql-jdbc.execute/do-with-connection-with-options
     :oracle
     (mt/db)
     {:session-timezone "US/Pacific"}
     (fn [^com.mchange.v2.c3p0.C3P0ProxyConnection proxy-conn]
       (let [^oracle.jdbc.OracleConnection conn (.unwrap proxy-conn oracle.jdbc.OracleConnection)
             t (oracle.sql.TIMESTAMPLTZ. conn "2024-06-20 17:28:00 -07:00")]
         (u.date/parse (.stringValue t conn)))))))
;; => #t "2024-06-20T17:28-07:00[US/Pacific]"
```

Reply:
Returning a ZonedDateTime here is pretty nice so maybe we should actually do this for `TIMESTAMPTZ` as well

Reply:
Nice to keep information if we have it

Reply:
Cool I'm on board with parsing strings. I chose to use `.getNString` instead of `.stringValue` because it means we can avoid unwrapping the connection, and I can't imagine why it would be less reliable.